bindings/profilers/wall.hh

```diff
@@ -2,28 +2,78 @@
 
 #include <nan.h>
 #include <v8-profiler.h>
+#include <unordered_map>
+
+#include "../buffer.hh"
 
 namespace dd {
 
+using LabelSetsByNode = std::unordered_map<const v8::CpuProfileNode*, v8::Local<v8::Array>>;
+
 class WallProfiler : public Nan::ObjectWrap {
  private:
   int samplingInterval = 0;
   v8::CpuProfiler* cpuProfiler = nullptr;
+  // TODO: Investigate use of v8::Persistent instead of shared_ptr<Global> to
+  // avoid heap allocation. Need to figure out the right move/copy semantics in
+  // and out of the ring buffer.
+  std::shared_ptr<v8::Global<v8::Value>> labels_;
+  bool labelsCaptured = false;
+
+  struct SampleContext {
+    std::shared_ptr<v8::Global<v8::Value>> labels;
+    uint64_t timestamp;
+
+    // Needed to initialize ring buffer elements
+    SampleContext() {}
+
+    SampleContext(std::shared_ptr<v8::Global<v8::Value>> l, uint64_t t)
+        : labels(l), timestamp(t) {}
```

Comment:
Hm… might it actually be even better to declare the argument as a `shared_ptr&`? Then it wouldn't be copied on the call site but would otherwise be fed correctly to the `labels` copy constructor here.